### PR TITLE
Feat/search ux

### PR DESCRIPTION
**Pull request by orthagh**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

image

### CODE REVIEW

**Comment by cconard96** [comment]
Here is a commit that fixes some JS conflicts after the search results are refreshed via AJAX and also fixes an issue where the criteria disappears after a search.
https://github.com/glpi-project/glpi/commit/69ab7d5b877ee9b27be9993adac549948688d9bd

**Comment by cconard96** [comment]
I made a few fixes on a fork of this PR:
https://github.com/cconard96/glpi/commits/feat/search_ux

The only remaining issue I noticed is when refreshing the results, the `ajax/search.php` endpoint occasionally responds with a JS script tag that would trigger a redirect to the saved search I selected last instead of the results data. This is considered an unexpected response by the Fluid Search JS, so it triggers a full page reload. After the reload, the previously selected saved search is showing in the button label even though I have custom criteria selected.

**Comment by orthagh** [comment]
TODO:

- [ ] savedsearches split
- [ ] responsive (hide some labels)

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
1) The thrash bin toggle is lacking some padding and probably the same rounded borders that are used by the display type toggle:

image

2) You can also see in the previous image that the map toggle is missing when inside the thrash bin.

3) Didn't you also wanted to integrate the kanban here ? Maybe delayed to another PR ?

image

4) The "filter" dropdown is not highlighted on my side compared to your screen shot.
Maybe it should have some hint of the primary color as it is an important action (more important than sorting IMO) ?

image

5) When using the map mode, the filter indicate that items are restricted on latitude and longitude but we can't see the associated criteria.

image

Maybe it will confuse some users ? Maybe the lat and long criteria should be shown as readonly ?

6) After using massive actions the items remained checked when the page is reloaded but the "actions" button is not displayed.

image

----

Otherwise I think it overall looks great, very clean 👍

**Comment by orthagh** [comment]
- 1, 4 & 6 fixed
- 2 cannot replicate on my side
- 3 later

**Comment by orthagh** [comment]
2 fixed

**Review by AdrienClairembault — APPROVED** [review]

File: src/Search.php [hunk 1]
```diff
@@ -378,6 +378,22 @@ public static function showGenericSearch($itemtype, array $params)
...
+     * Print generic sort form
+     *
+     * Params need to parsed before using Search::manageParams function
+     *
+     * @param class-string<CommonDBTM> $itemtype  Type to display the form
+     * @param array  $params    Array of parameters may include sort, order
+     *
+     * @return void
+     **/
+    public static function showGenericSort($itemtype, array $params)
```
Comment (cconard96) [inline src/Search.php:392]: Not sure these two methods need added here. `Search` only exists still for backwards-compatibility.

File: src/Search.php [hunk 2]
```diff
@@ -461,6 +477,24 @@ public static function displaySearchoptionValue($request = [])
     }
 
 
+    /**
+     * Display a criteria field set, this function should be called by ajax/search.php
+     *
+     * @since 9.4
```
Comment (cconard96) [inline src/Search.php:483]: Wrong version

**Review by cconard96 — CHANGES_REQUESTED** [review]

File: src/Search.php (hunk 1 above)
Comment (cconard96) [reply src/Search.php:392]: I think the warnings on `Glpi\Search\SearchEngine` and `Search` can be removed/ignored now too.

**Comment by trasher** [comment]
Search seems OK, I had an issue with sorting. I've selected:
image

And once I've applied sort, this is shown:
image


I can see following in logs:
```
[2023-11-29 07:51:49] glpiphplog.WARNING:   *** PHP Warning (2): Undefined array key "" in /var/www/webapps/glpi/src/Search/Output/HTMLSearchOutput.php at line 167
  Backtrace :
  src/Search.php:350                                 Glpi\Search\Output\HTMLSearchOutput::displayData()
  ajax/search.php:90                                 Search::displayData()
  public/index.php:82                                require()
  
[2023-11-29 07:51:49] glpiphplog.WARNING:   *** PHP Warning (2): Trying to access array offset on value of type null in /var/www/webapps/glpi/src/Search/Output/HTMLSearchOutput.php at line 167
  Backtrace :
  src/Search.php:350                                 Glpi\Search\Output\HTMLSearchOutput::displayData()
  ajax/search.php:90                                 Search::displayData()
  public/index.php:82                                require()
```

**Comment by trasher** [comment]
I guess this happens when I sort on a non displayed column

**Comment by AdrienClairembault** [comment]
I guess it doesn't make sense to be able to sort multiple time on the same column: 

image

File: js/modules/Search/Table.js [hunk 3]
```diff
@@ -151,6 +153,35 @@ window.GLPI.Search.Table = class Table extends GenericView {
...
+        $(sort_container).find('select[name^="sort"]').each(function() {
+            sorts.push($(this).val());
+        });
+
+        // Remove data-sort-num from all headers and change all data-sort-order to nosort to reset the sort state
+        this.getElement().find('thead th').each((i, c) => {
+            const col = $(c);
+            col.attr('data-sort-num', null);
+            col.attr('data-sort-order', 'nosort');
+        });
```
Comment (AdrienClairembault) [inline js/modules/Search/Table.js:172]: FYI jQuery's `attr` method work on the whole collection, there is no need to iterate on each items.

Something like that would work for example:

```suggestion
        this.getElement().find('thead th').attr({
            'data-sort-num': null,
            'data-sort-order': 'nosort',
        });
```

**Comment by AdrienClairembault** [comment]
> > I guess it doesn't make sense to be able to sort multiple time on the same column:
> 
> Right, but there's more in fact, I just think about it : but as we defer the sort from the selects to th, any options not present in table header will not be sorted.
> 
> @cconard96 I guess we need to track sorts/orders in a class property and both sources of sorting should alter them rather than forwarding to headers.

A temporary fix would be to allow only the displayed headers to be selectable in the sort list.

**Comment by orthagh** [comment]
Sort/order issues should be fixed.
More, I added "used" management, adding a new sort row (from card panel), should now take into account previous sorted columns and not presented them

**Review by cconard96 — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**